### ISSUE TITLE
Better dead code elimination

### ISSUE THREAD

**Issue by kangax** [eslint/eslint#6255]
no-unreachable fails to detect some of the cases of unreachable code. For example,

```js
(function() {
  var foo = null;
  if (!foo) {
    return;
  } 
  (function dead(){})();
})();
```

should warn about `dead` function (since it's statically determinable).

**Comment by ilyavolodin** [comment]
While this is statically determinable to the point, we usually stayed away from trying to inspect values of the variables. This is more of the realm of Flow. I'm not against adding something like this in, but historically this is something we stayed away from.

**Comment by kangax** [comment]
I understand that it's a bit of a rabbit hole. Perhaps separate (experimental?) rule? Good point about Flow; unfortunately it's just not always possible to use it (even though we at Facebook try to, there's still older files that's difficult to port or that no one just got around to).

**Comment by mysticatea** [comment]
Unfortunately, ESLint doesn't have type inference feature.
Though it's one of what I want to do.

**Comment by nzakas** [comment]
Historically we've stayed away because we just didn't have any easy way to do this.

@kangax can you share how you were able to identify dead code in your project? Did you end up needing Flow information or were you doing something else?

**Comment by kittens** [comment]
We're working on some Babel tooling that uses it's APIs. Our scope tracking tracks constant values and knows that `foo` will always be `null`. Then some of our evaluation helpers determine that `if (!foo)` will always be `true`. There's no Flow type information required.

**Comment by ilyavolodin** [comment]
If we were to add something like this to ESLint, I would seriously consider adding it to Escope instead. Escope already checked all of the declarations in the AST, so enhancing it to track type inference would be simpler then adding something like that to ESLint core itself.

**Comment by mysticatea** [comment]
A week ago, I made [a rule which uses value tracking](https://github.com/mysticatea/eslint-plugin-node/commit/624aed0ead4aa1fd2a656de4c6ca5ca3e21ccb99). This rule got pretty complex, so it's great if type inference / value tracking come to ESLint core (or escope).

**Comment by nzakas** [comment]
@kittens Thanks for the info, that's very helpful.

Agree with @ilyavolodin, if this was going to be supported, we'd want it in escope rather than ESLint core. If they aren't willing to do that, then we'd probably need to fork escope to add it in.

For now, I'd say this is a pretty low priority until we get through the JSCS work. So we can either leave this open and not worry about working on it or we can close this and, at some point in the future, talk more about adding some basic type inference.

**Comment by michaelficarra** [comment]
@nzakas @ilyavolodin I wouldn't say this is appropriate for escope at all. Currently, escope knows nothing about control flow or about JavaScript's values. This would be a completely unrelated responsibility, and if it was proposed to be added to escope, I would strongly object to it.

**Comment by ilyavolodin** [comment]
@michaelficarra I think I might've phrased it wrong. What I meant was that escope is a better fit for this then ESLint core. It might be not be the best fit, and we might need to create another tool to deal with this, but I don't think it should be part of ESLint core.

**Comment by kangax** [comment]
@ilyavolodin if not part of ESLint core, then the other option is to just make a standalone (unofficial) plugin?

**Comment by ilyavolodin** [comment]
@kangax That would be one way to do this, but it would be very complex. ESLint uses Escope to track variable scoping. It also uses estraverse to navigate AST. As I mentioned above, solution might be to enhance escope to not jus track variable scoping, but also variable typing. Or if that's not an option, creating another tool that basically mimics what Flex does and plugging it into ESLint core would also work.

**Comment by nzakas** [comment]
@michaelficarra escope does most of what is needed already. We'd just need to attach type information when a declaration is found. The scope, type, and evaluation of identifiers are pretty closely related, so I don't see a good argument for having multiple passes from different tools to do this.

In any event, I don't see this as something that will be addressed anytime soon, so I'd propose we close this issue.

**Comment by sheerun** [comment]
Relevant: https://www.youtube.com/watch?v=VEaDsKyDxkY

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
